### ISSUE TITLE
Update connectionPayload for subscription on running Apollo instance (for Bearer-Token).

### ISSUE THREAD

**Issue by schoeda** [apollographql/apollo-kotlin#3666]
We use in our Android client wss with the Apollo-GraphQL SDK. To subscribe to GraphQL we add a `connectionPayload`-Map of the `SubscriptionWsProtocol` with a Bearer-Token. If the Bearer get invalid, we have to recreate the whole Apollo client and add all subscriptions again.

Solution:
Make it possible to update the `connectionPayload` at runtime, without a recreation of the Apollo client.

Thank you!

**Comment by martinbonnin** [comment]
Hi again 👋 

How does your server signals an expired `Bearer-Token`? We might add something like this but AFAIK, there's no real standard way to do it like 401 errors with HTTP so having a sample of what the exchange would look like would help a ton!

**Comment by schoeda** [comment]
Hey again, Martin :) 

I tested this only with an invalid Token at app start and we get an error in the subscription flow.

```
apolloSyncClient.subscription(ChangedFacilityPickJobsSubscription()).toFlow().collect { subscriptionResult ->
    if (subscriptionResult.hasErrors()) {
        // Errorhandling
    } else {
        // Do funny things...
    )
}
```

The error looks like this:

```
Error(message = Unauthorized, locations = null, path=null, extensions = {code=UNAUTHENTICATED, response={statusCode=401, message=Unauthorized}}, nonStandardFields = null)
```

I know, this is a little bit different than an expired token. But if that happens, we create a complete new instance of the Apollo client.

**Comment by martinbonnin** [comment]
Tentative API there: https://github.com/apollographql/apollo-android/pull/3674

Because `connectionPayload` is a lambda, you should be able to provide a new value when needed. And use `reconnectWhen {}` to decide whether to reconnect or not. Let me know how that looks!

**Comment by matejcikm** [comment]
Hi @martinbonnin, is possible to implement `reconnect()` method in `WebSocketNetworkTransport`? I think it will be helpful for handle bearer tokens expiration errors explained above. `reconnectWhen` lambda is not ideal for this situation, because it is called only after `NetworkError` (disconnecting from webSocket, etc...).

**Comment by martinbonnin** [comment]
Hi @matejcikm 👋 

The problem here is that I'm not sure there is a standard way to refresh tokens so everything we do should be generic enough to work with multiple protocols (think AppSync, `graphql-ws` and other protocols). We'd need a precise message to trigger the refresh. Would you have a dump of what happens in that case? (Charles proxy dump or Chrome webtools or anything like this?)

> reconnectWhen lambda is not ideal for this situation, because it is called only after NetworkError (disconnecting from webSocket, etc...).

Just making sure I understand, is the worry that you have to setup a new TCP connection and negociate a new WebSocket upgrade? This is indeed sub-optimal but I would not expect this to be a huge issue in practice? Or am I missing something?

**Comment by SumeraMartin** [comment]
@martinbonnin hi, I know there is no standard way of dealing with refresh tokens. So it means that it should be flexible enough to cover them all :-)

I would like to add a sample when the reconnection with `webSocketReconnectWhen` [#3674](https://github.com/apollographql/apollo-android/pull/3674)  is not possible.

Our graphql response doesn't return any HTTP error. It just returns a response with `errors`.

```
{
    "data": "",
    "errors": ["Unathorized error"],
 }
```

This won't produce an error so there won't be `NetworkError` and thus the following block won't be executed. Only the "normal" subscription message will be received. In this case, there is no way to trigger reconnect operation.

<img src=https://user-images.githubusercontent.com/6089152/145841814-aa40c417-835f-419d-b7a7-540e9bdb6af7.png width=550 />


I'm not sure what is the best solution for this problem but I can imagine that there could `webSocketReconnectWhen` override with a received data message instead of an error. Another solution could be to have a reconnect method implemented as mentioned by @matejcikm In the previous version of apollo this was possible via `SubscriptionManager ` but that doesn't exist anymore.

**Comment by martinbonnin** [comment]
Thanks for sending this 👍 . That helps a ton. I'd like to see cases like these be handled at the `WsProtocol` level, maybe passing a custom instance or so we don't even have to tear down the WebSocket. I'll share a tentative API later today.

**Comment by martinbonnin** [comment]
Hi 👋 . It turns out it's possible to intercept messages and close the WebSocket on error messages using the current APIs. See the [test](https://github.com/apollographql/apollo-android/pull/3697/files#diff-bbefdc8d0f60d9b9281adf7622146aaaf29dba71dab915ddbce82720bd9badb7R230) in #3697 

```kotlin
  private object AuthorizationException: Exception()

  private class AuthorizationAwareWsProtocol(webSocketConnection: WebSocketConnection, listener: Listener) : WsProtocol(webSocketConnection, listener) {
    val delegate = SubscriptionWsProtocol(webSocketConnection, listener)

    override fun handleServerMessage(messageMap: Map<String, Any?>) {
      val isError = // checkMessageMap for an error
      if (isError) {
        /**
         * The server returned a message with an error and no data. Send a general error upstream
         * so that the WebSocket is restarted
         */
        listener.networkError(AuthorizationException)
      } else {
        delegate.handleServerMessage(messageMap)
      }
    }

    override suspend fun connectionInit() {
      delegate.connectionInit()
    }

    override fun <D : Operation.Data> startOperation(request: ApolloRequest<D>) {
      delegate.startOperation(request)
    }

    override fun <D : Operation.Data> stopOperation(request: ApolloRequest<D>) {
      delegate.startOperation(request)
    }
  }

  class AuthorizationAwareWsProtocolFactory: WsProtocol.Factory {
    override val name: String
      get() = "graphql-ws"

    override fun create(webSocketConnection: WebSocketConnection, listener: WsProtocol.Listener, scope: CoroutineScope): WsProtocol {
      return AuthorizationAwareWsProtocol(webSocketConnection, listener)
    }
  }

  @Test
  fun canResumeAfterGraphQLError() {
    val wsFactory = AuthorizationAwareWsProtocolFactory()
    val apolloClient = ApolloClient.Builder()
        .serverUrl("http://localhost:8080/subscriptions")
        .wsProtocol(wsFactory)
        .webSocketReconnectWhen {
          it is AuthorizationException
        }
        .build()
```

I can see how that's verbose though so we could think of more concise APIs. 

To build on the 2.x idea of being able to "reconnect" at any point in time, what about something like this:

```kotlin
/**
  * closes the currentConnection if any is open. All listening `Flows` will throw an error and be terminated
  * Does nothing if no connection is currently open
  * 
  * @param error: the exception to throw. Use this in combination with `reconnectWhen` to either teardown all 
  * `Flow` or resubscribe them automatically once the connection is resumed
  **/
fun WebSocketNetworkTransport.closeCurrentConnection(error: Exception) {
  ...
}
```

**Comment by CureleaAndrei** [comment]
Hi, the mechanism you described, build on the 2.x ideea where you can close the connection and the reconnect would be very use-full. Is there any progress on this ? Thanks.

**Comment by BoD** [comment]
Hi @CureleaAndrei !

We're looking into implementing it but in the meantime can you tell us a bit about your use-case? Do you also need this to handle expired (or otherwise changed) auth tokens, or something else? Are you in a multi-user app? Understanding the usage would help us defining a robust API.

**Comment by CureleaAndrei** [comment]
Hi, my use case is the following: I get the JWT from firebase, which is an async operation. The app supports endpoints that needs authorisation (we have a permission sistem on the server that returns a  response with a custom “Forbidden” error, not a HTTP error) and endpoints that don’t need authorisation and are public for all users. Once I register these subscription I need to update the payload in an async way without closing/completing the already made subscriptions(flows). Please note that the custom ‘Forbidden’ is not a token expiration so we don’t need to retry the connection. We just need to handle it and display it accordingly in app.

The moments when I want to update in an async way the payloads are : when the user changes and when the token expires.  One problem encountered here is that at the app start-up we don’t have yet a token and the ApolloClient is initialised without a JWT, the subscriptions are called, we get the forbidden, and then the JWT comes. Recreating here the ApolloClient breaks all the subscriptions. Same happens when the user log-out.  I managed to avoid subscription breaking by doing a workaround: 
Having a flow that fires at user change and: 
 1) Recreate ApolloClient with the new jwt
2) FlatMapping the user change flow and returning the apollo client subscription

For token expiration we think of sending an other custom error like “Expired” or “Invalid” token that we need to handle and retry the connection with a new token.  For this case a mechanism like SubscriptionWsProtocolAdapter where we can intercept errors from response and throw them to be handled in retryWhen{} mechanism it would be awesome. 

I played a little with SubscriptionWsProtocolAdapter but I think there is something missing, or I haven’t figured it out. Using the SubscriptionWsProtocolAdapter I haven’t found a way to customise (like setting the connectionPayload, etc.) the underlying delegate SubscriptionWsProtocol because it’s private. Also this mechanism should be made for GraphQLWsProtocol, or maybe a generic Protocol Adapter that can be customised with the desired protocol and it’s settings.

 Wrapping it up it would be awesome to have a mechanism that triggers from outside a reconnection (throwing an error from outside that would be handled in reconnectWhen like you mentioned seems like a healthy approach) and a mechanism that triggers the reconnection based on the response from the server (same like above, being able to throw an error and also handling it in the reconnectWhen would be awesome. The ProtocolAdapter with some upgrades mentioned above I think handles this case pretty well). Thanks and hope this feedback is use-full.

**Comment by martinbonnin** [comment]
Hi! Thanks for all the details! This helps a ton 💙 . Look like `closeCurrentConnection(exception)` is the way to go then (or maybe just use `reconnect(exception)` to match the 2.x naming?). 

For the sake of completeness, do you mind elaborating a bit on this part:

>  at the app start-up we don’t have yet a token and the ApolloClient is initialised without a JWT, the subscriptions are called, we get the forbidden, and then the JWT comes

How do you get the `Forbidden`? Is it a message on the websocket or something out of band? 

> the custom ‘Forbidden’ is not a token expiration so we don’t need to retry the connection. We just need to handle it and display it accordingly in app

Looks like the logic to retrieve/update your JWT token is decoupled from the subscription? What is triggering the JWT update then? 
Reason I'm asking is that if it is something like a signin and you're using the cache, you will also want to clear the cache or you'll have stale results and the safest way to do this is to re-create the ApolloClient.

**Comment by CureleaAndrei** [comment]
In my opinion closeCurrentConnection(exception) is better than to use the old reconnect(exception) naming if we need to handle the error in reopenWhen{}. If it just reconnect without throwing an error then the reconnect naming from v2 is better.

Giving the jwt token in a sync way at the creation time of ApolloClient means that I have to block the thread until the JWT token comes from network, because the token retrieval is async. To avoid that I create the ApolloClient with no jwt. 

We have 2 cases when we need to update the payload: 
1. When the user changes (Logs out/Signs in)
2. When the token expires

For:
1. The trigger for the recreation of ApolloClient is a flow that fires anytime when the user changes (ex. : Logs out or signs in) and in which we retrieve the JWT for the new user and then recreate the ApolloClient. The disadvantage here is that we lose the already made subscriptions. This is why the closeConnection(exception) and the handling in reopenWhen{} would be handy.
2. We will handle it using the ProtocolAdapter and reopenWhen{} mechanism

The "Forbidden" is an error that we send in response (where we have data and errors).

And yeah, you are right, clearing the cache is also necessary when the user changes (case 1).

Thanks for doing it. If I can provide any more informations, I will gladly help or If I understood something wrong let me know.

**Comment by BoD** [comment]
@CureleaAndrei `closeConnection()` has been released in v3.3.0!  I'm closing this issue for now but don't hesitate to let us know if that helps with your use-case, and to re-open an issue if more is needed.
